### PR TITLE
fix(interactives): number conditional blocks in sections

### LINKED ISSUE #1601: In-Section Conditional Blocks are not numbered
Using a conditional block within a section does not create a numbered list item.Example:

Image

Image

### PR DESCRIPTION
**Pull request by dvd233**
## Summary

Conditional blocks inside sections were omitted from the visible sequence because `InteractiveConditional` opted out of section numbering. This change makes a conditional wrapper occupy one authored position and aligns inline passive or media branch content with step cards.

An initially empty selected branch — including parsed elements that produce no React output or nested conditionals that settle empty — consumes no counter. Once a populated conditional has appeared, a later live re-evaluation to an empty branch retains an invisible, zero-height counter slot so the numbers already shown for following steps do not change. Empty nested conditionals stay mounted so their own requirement listeners can recover; visible siblings remain intact, and section-display conditionals keep their nested section counter independent.

## What to look at

- [Section numbering](https://github.com/dvd233/grafana-pathfinder-app/blob/71f0380c1ad521df9d95f3975ab0ffe6cf8b37d8/src/components/interactive-tutorial/section-numbering.tsx) treats a conditional wrapper as one outer position while continuing to exclude media.
- [Conditional rendering](https://github.com/dvd233/grafana-pathfinder-app/blob/71f0380c1ad521df9d95f3975ab0ffe6cf8b37d8/src/components/interactive-tutorial/interactive-conditional.tsx) checks direct React output, observes committed nested DOM recursively, distinguishes pending loading from settled emptiness, and projects empty/retained state onto a wrapper that keeps nested listeners mounted.
- [Interactive styles](https://github.com/dvd233/grafana-pathfinder-app/blob/71f0380c1ad521df9d95f3975ab0ffe6cf8b37d8/src/styles/interactive.styles.ts) uses class-based `:has` selectors that are exercised in jsdom to suppress never-occupied empty wrappers and collapse retained slots without removing them from the CSS counter.

## How to verify

1. Render a section with a normal block, an inline conditional, and another normal block; confirm the visible numbers are consecutive and the conditional occupies its authored position.
2. Switch a conditional between passive HTML, media, and interactive-step branches; confirm each branch stays aligned under the same outer number.
3. Select an initially empty branch, including one with parsed children whose renderer returns no output, followed by another block; confirm the empty branch creates no visible number or gap.
4. Render conditionals nested at several levels; confirm an all-empty chain produces no number, then make an inner `exists-reftarget` pass without re-evaluating its parent and confirm the content becomes visible. Confirm an empty nested conditional beside visible content leaves its sibling and outer number intact.
5. After populated nested content is visible, re-evaluate the inner conditional to empty; confirm the mounted outer wrapper enters retained state and following step numbers do not change.
6. Render a section-display conditional; confirm it occupies one outer position while its nested section starts and advances its own counter independently.

## Follow-up

The latest follow-up keeps empty and retained numbering markers across condition re-evaluations, so a same-branch check cannot resurrect an orphan ordinal. It extends the original section-numbering contract rather than replacing it, and adds regression assertions for both empty and retained nested conditionals.

None. This change restores the intended section numbering and is fully reversible.

Fixes #1601

🤖 Generated with [OpenAI Codex](https://openai.com/codex/)

### CODE REVIEW

**Review by moxious — CHANGES_REQUESTED** [review]
Thanks for tackling section numbering for conditional blocks — removing the `skip-section-numbering.ts` marker mechanism in favor of letting `InteractiveConditional` own its own numbered slot is a nice simplification, and the new nested-conditional test coverage is appreciated.

I checked out the PR head locally, ran the full changed-area test suite (all 555 tests pass), and reproduced one behavior with a targeted test against your actual code before writing this up.

**F1 — blocking, medium severity.** `interactive-conditional.tsx` (~L265-267) / `interactive.styles.ts` (~L363-366): making the conditional itself numbered means an already-mounted, already-populated (and thus numbered) conditional can *live-collapse* after the user has already seen it, shifting every subsequent step's visible number. Any of the conditional's existing re-evaluation triggers (`interactive-action-completed`, `datasources-changed`, `plugins-changed`, `popstate`, the `exists-reftarget` MutationObserver) can flip a populated branch to empty post-mount, and the new `:empty { display: none }` rule then drops it from the CSS counter live, in front of the user. I reproduced this: mounted a conditional with `pass: true` (numbered, populated), then dispatched `interactive-action-completed` with the check flipped to `pass: false` — the wrapping `<li data-numbered="true">` goes from occupied to `display: none` post-mount. Pre-PR, `InteractiveConditional` never received a number at all, so this class of mid-session renumbering didn't exist. Only the mount-time collapse path is covered by the PR's own test; the live/post-mount case isn't. Could you confirm whether this is an accepted tradeoff (if so, a regression test asserting it explicitly would help future readers) or whether it should be avoided — e.g. not letting an item leave the shared count once it has entered it, or only applying the `:empty` collapse before first paint?

**F2 — follow-up, low-medium severity.** `interactive-conditional.tsx` (~L28-34): `markPlainNumberingChild` only tags children where `typeof child.type === 'string'` (plain DOM elements). A branch child that renders as `ImageRenderer`/`VideoRenderer`/`YouTubeVideoRenderer`/`VimeoVideoRenderer` (all function components, and all valid inline-branch children per `content-renderer.tsx`) is skipped, so it never gets the `section-numbering-plain` class or the new alignment padding — it'll sit flush against the `<li>` edge while sibling text aligns correctly. No test exercises an image/video branch child; worth extending the tagging to cover the same identity list `shouldNumberSectionChild` already uses.

**F3 — nit.** `interactive-conditional.tsx` (~L265-267, 296): the `section-numbering-plain` tagging runs unconditionally even in `display="section"` mode, where `InteractiveSection` wraps its own children via its own numbering pass — so the tagged element is never a direct child of `.interactive-conditional` there and the CSS selector can't match. Harmless, but it's dead work on a path where it can never fire; a short comment (or skipping it for that branch) would save a future reader some head-scratching.

**F4 — nit.** No test directly asserts that two numbered conditionals (or a conditional plus a real interactive step) sitting side-by-side keep a correct shared `1, 2, 3…` sequence — the new tests all number-check within a single conditional's own subtree. The CSS counter mechanism itself is unchanged, so this is likely fine, but it's a cheap, valuable regression test given this PR is specifically about that sequence.

Requesting changes mainly on F1 — want to make sure the live-renumbering case is a deliberate decision rather than a side effect before this merges. Happy to take another look once that's addressed.

**Comment by dvd233** [comment]
@moxious Thanks for the detailed review. F1–F4 are addressed in 34e9e2bf: live collapse now retains an invisible zero-height counter slot after a populated branch has appeared; known media renderers share the inline alignment path; section mode skips the unused tagging; and the suite now covers live collapse, media identity, section mode, and adjacent numbering. Local verification passed the targeted 39 tests, the full interactive-tutorial area (30 suites / 531 tests), typecheck, repository ESLint, and Prettier. The PR description has also been updated to document the resulting semantics. Ready for another look.

File: src/components/interactive-tutorial/interactive-conditional.tsx [hunk 1]
```diff
@@ -281,19 +268,17 @@ export function InteractiveConditional({
...
   }
 
-  // Select the appropriate branch and config based on condition result
   const childrenToRender = conditionsPassed ? whenTrueChildren : whenFalseChildren;
   const sectionConfig = conditionsPassed ? whenTrueSectionConfig : whenFalseSectionConfig;
 
-  // If the selected branch has no children, skip rendering entirely
   if (childrenToRender.length === 0) {
-    return null;
+    return hasOccupiedNumberingSlot ? <span hidden data-section-numbering-retained="true" /> : null;
```
Comment (moxious) [inline src/components/interactive-tutorial/interactive-conditional.tsx]: Blocking (conditional-empty-render-orphan-number): this guard is the only thing that suppresses the number, and it counts ParsedElements while the `:empty` rule that does the suppressing is DOM-based. A branch that holds elements which render to nothing still falls through to the inline wrapper `<div className="interactive-conditional ...">`, so the `<li>` is neither `:empty` nor holding the retained span, and the base rule at src/styles/interactive.styles.ts:385-401 paints an ordinal onto a zero-height row plus shifts every later step by one.

The shape that reaches it is a conditional nested in another conditional's branch: the inner one returns `null` when its own branch is empty and it has never been occupied. Branch arrays carry no `.min(1)`, the conditional arm recurses to MAX_NESTING_DEPTH 5, and no validator rejects an empty branch.

Suggestion: build `renderedChildren` first and, if every entry is `null` or `false`, return the same result this guard returns.

**Review by moxious — CHANGES_REQUESTED** [review]
Thanks for the second round on conditional numbering — this one clears all four findings from the last review, including the live-collapse renumbering that was the previous blocker. The retained-slot latch is a neat answer to it. One new blocker below, in the same area.

---

## Merge contract

Fix this item and this PR is mergeable.

1. [blocking] **conditional-empty-render-orphan-number — A conditional branch that renders nothing leaves a stray step number and shifts the rest of the section** (high · correctness-and-reliability)
   The empty-branch guard tests `childrenToRender.length === 0`, which counts ParsedElements, while the new `:empty` rule that suppresses the number is DOM-based. When a branch holds elements that render to nothing, the inline path still emits its `<div className="interactive-conditional ...">` wrapper, so the `<li>` is not `:empty` and does not carry the retained-slot span either. Neither new rule fires, and the base rule at src/styles/interactive.styles.ts:385-401 still applies `counterIncrement` plus an absolutely positioned `::before`. `.interactive-conditional` has no styles outside `.loading`, so the wrapper is a zero-height block and the `<li>` is `data-step="true"` with no paddingTop: the ordinal paints 16px down, onto the following item's own ordinal, and the consumed counter leaves every later step in the section one higher than its authored position. The reachable shape is a conditional nested in another conditional's branch whose inner branch is empty on first evaluation, so the inner one returns null. Branch arrays carry no `.min(1)`, the conditional arm recurses to MAX_NESTING_DEPTH 5, and no validator rejects an empty branch. A milder second shape is a branch whose only child is a markdown block that parses to nothing. On the base commit the wrapper `<li>` carried no `data-numbered` attribute at all, so no ordinal could paint.
   Required: Decide emptiness from rendered output rather than the ParsedElement count: after building `renderedChildren`, if every entry is null or false, return the same null / retained-span result the length-zero guard already returns. A CSS-only alternative is a companion rule for an empty `.interactive-conditional` wrapper. Please cover the nested-conditional shape with a test either way.

## Follow-ups

These are tracked separately and do not block merge.

1. [follow_up] **retained-slot-collapse-unverified — The retained-slot collapse ships with no assertion that can fail** (medium · testing-and-verification)
   The mechanism that resolves the prior round's blocker is the `:has(> [data-section-numbering-retained="true"])` rule, and it is the one part of this change no test can hold. Probing this repo's own environment (jsdom 26.1.0, nwsapi 2.2.24) shows nwsapi cannot parse an unprefixed attribute selector after a child combinator inside `:has()`, so that rule is silently skipped in the cascade: `getComputedStyle` returns empty strings for height, overflow and padding, byte-identical to a run with the rule deleted, with no throw and no test failure. `getComputedStyle(el, '::before')` is unimplemented, so the `::before { display: none }` half is unassertable by any means. The two assertions offered in their place are already covered by their neighbours: `display).not.toBe('none')` fails only if the retained span were absent, which the preceding querySelector line asserts, and `counterIncrement).toBe('step-counter')` comes from the pre-existing base rule and fails only if data-numbered were absent, which the next line asserts. Without the collapse rule the retained slot renders as a full-height numbered row with nothing beside the ordinal, which is exactly what the mechanism exists to prevent.
   Follow-up: Restate the selector in the descendant form nwsapi parses, `li[data-numbered="true"]:has([data-section-numbering-retained="true"])`, which does cascade in this jsdom, and then assert height, overflow and padding on the retained item. Alternatively drop the two decorative assertions and record in a comment that the collapse is browser-only and was checked by hand.

## Suggestions

1. [suggestion] **numbering-stability-contract-unstated — The numbering-stability trade-off is real in three directions and stated in none of them** (medium · correctness-and-reliability)
   Numbering a wrapper makes the visible sequence move in three places, and none of them is written down. First, while checking, the conditional renders the 80px bordered `.interactive-conditional.loading` box, which now occupies a counter slot; when it resolves to an empty branch the `<li>` becomes `:empty`, `display: none` releases the counter, and every following step decrements. Some conditions do real HTTP (src/requirements-manager/checks/grafana-api.ts:383 and :467), so that window is paintable, and the PR body's claim that an initially empty branch consumes no counter holds only for the settled state. Second, the retained slot hides its own `::before` while keeping `counterIncrement`, so after a live collapse the sequence has a hole and a section can read 2. 3. with no 1.; that is the accepted cost of the fix this reviewer asked for last round, not a new objection. Third, nothing reserves a slot before first occupancy, so a conditional that first resolves empty and later flips to a populated branch inserts a number mid-list and renumbers upward, and that path has no test while the other two each have one.
   Suggested: Write the rule down next to the mechanism and correct the PR body: a slot is reserved from first occupancy onward, a reserved-but-empty slot shows as a gap, and the mount-time and empty-to-occupied transitions do renumber. If the mount-time shift is unwanted, the alternative is to leave an unresolved conditional out of the count so its number arrives with its content rather than ahead of it.

2. [suggestion] **conditional-numbering-test-fidelity — The new alignment tests are brittle and stop short of the property they name** (low · testing-and-verification)
   The alignment assertions hard-code '16px' and 'calc(18px)' in five places; both come from `createTheme().spacing.gridSize === 8`, so a theme change breaks them with a message about pixels rather than spacing. In `keeps passive and interactive branches aligned under one number` the interactive branch is a hand-rolled `<div className="interactive-step">` and the assertion is `paddingLeft).not.toBe('calc(18px)')`, which also passes for a bare unstyled div: it catches over-marking but never checks that the two branches land in the same column. `aligns a passive leaf through nested conditionals without compounding the offset` asserts paddingLeft on the leaf only, while compounding would arrive through ancestor padding. The media case covers ImageRenderer alone, and VideoRenderer accepts className only implicitly through its index signature and `{...props}` spread, so a future destructure there would drop the alignment class with nothing failing. The file also hand-rolls mocks for `../../interactive-engine`, `../../requirements-manager` and `../../constants` that src/test-utils/interactive-section-harness.tsx already provides, and the requirements-manager one is a total replacement with no `jest.requireActual`, so a later import into interactive-section.tsx becomes undefined in this file alone.
   Suggested: Derive the expected values from one `createTheme()` rather than pixel literals, either rename the interactive-branch assertions to what they check or assert the two column positions as an equality, extend the media case to an it.each over all four renderers, and take the three module mocks from the shared harness.



---

Next step: the one blocking item is the empty-render case — the guard counts parsed elements, but the CSS that suppresses the number is DOM-based, so a branch that renders to nothing slips between them. The two follow-ups and the two suggestions do not need to land here. CI is green on this head (the `publish-report` and Bundle Stats failures are a known fork-permission limitation on our side, not yours). Push a fix and CI will rerun; ping me and I will re-review.

**Comment by dvd233** [comment]
@moxious Thanks for the precise review. `7e796877` addresses the empty-render blocker:

- `InteractiveConditional` now checks the actual `renderElement` output before emitting its numbered wrapper, and the retained-slot latch records only content that really rendered; a parsed child that returns `null`/no output no longer consumes a number.
- A nested empty conditional wrapper is hidden only when it is the sole child, while a visible sibling remains numbered and visible.
- The existing populated-then-empty retained-slot behavior and section-display numbering remain unchanged.

Verification completed: the full `src/components/interactive-tutorial` area (30 suites, 533 tests), typecheck, full lint, and Prettier. Ready for another look.

File: src/components/interactive-tutorial/interactive-conditional.tsx [hunk 2]
```diff
@@ -17,39 +7,53 @@ import { isValidRequirement } from '../../types/requirements.types';
...
+    return node.some(hasRenderableNode);
+  }
+  if (typeof node === 'string') {
+    return node.length > 0;
+  }
+  if (React.isValidElement<{ children?: React.ReactNode }>(node) && node.type === React.Fragment) {
+    return hasRenderableNode(node.props.children);
+  }
+  return true;
+}
```
[the remark this answers is not in the document]
Comment (dvd233) [reply src/components/interactive-tutorial/interactive-conditional.tsx:55]: Thanks — this is valid. `9b2ac67c` replaces the component-element heuristic and one-level CSS selector with committed DOM-subtree observation. Nested conditional loading is treated as pending, hidden retained markers are not counted as visible parent content, and only actual DOM output latches occupancy. An all-empty chain now collapses at three levels, a visible sibling prevents collapse, and a populated nested branch promotes a direct retained marker when it later empties.

Verification completed: 30 interactive-tutorial suites / 534 tests, typecheck, full lint, and Prettier.

File: src/components/interactive-tutorial/interactive-conditional.tsx [hunk 3]
```diff
@@ -281,19 +370,19 @@ export function InteractiveConditional({
...
-  // If the selected branch has no children, skip rendering entirely
   if (childrenToRender.length === 0) {
-    return null;
+    return hasOccupiedNumberingSlot ? <span hidden data-section-numbering-retained="true" /> : null;
+  }
+
+  // Parsed elements can produce no React output directly, or a nested
+  // component can settle empty after commit. Neither may leave a numbered
+  // wrapper behind; only a slot that previously rendered stays reserved.
+  if (!hasRenderableChildren || isEmptyAfterCommit) {
```
[the remark this answers is not in the document]
Comment (dvd233) [reply src/components/interactive-tutorial/interactive-conditional.tsx]: Thanks — this is valid. `71f0380c` keeps nested conditional instances mounted when their committed DOM is empty. The wrapper now carries either `section-numbering-empty` or `section-numbering-retained`; the outer DOM observer stays active, and only the parent `<li>` visibility/collapse changes.

The regression suite now proves that an inner `exists-reftarget` can recover after a target is injected while the parent condition never re-evaluates, that three empty nesting levels remain mounted without consuming a number, and that a previously populated nested branch retains its outer slot. The retained selector is class-based and ordered after the base padding rules, so jsdom now asserts height, overflow, and padding rather than silently skipping the rule.

Verification completed: 30 interactive-tutorial suites / 535 tests, typecheck, full lint, and Prettier.

**Review by moxious — CHANGES_REQUESTED** [review]
Round 3. Five things from the earlier rounds are resolved and I have moved them to cleared state: the live-collapse renumbering, the media-branch tagging, the dead tagging in `display="section"`, the missing adjacent-conditional test, and the retained-slot collapse — that last one I verified by probing this repo's own jsdom/nwsapi rather than trusting the selector, and the old attribute form genuinely throws where the new class form matches. The `:has(> .section-numbering-retained)` change was the right call.

One item remains, and it is the round-2 blocker rather than a new one: the fix works on the way in and comes undone on the next re-evaluation. It keeps its original ID for that reason. Nothing else below — no nits, no follow-ups.

## Merge contract

Fix this item and this PR is mergeable.

1. [blocking] **conditional-empty-render-orphan-number — A re-evaluation clears the empty and retained numbering state, and nothing puts it back** (high · correctness-and-reliability)
   The new empty-render handling works on the way in and comes undone on the next re-evaluation. Both `section-numbering-empty` and `section-numbering-retained` are derived only from `isEmptyAfterCommit`, the identity test `emptyRenderToken === renderToken` at src/components/interactive-tutorial/interactive-conditional.tsx:221. `evaluateConditions` then calls `setEmptyRenderToken(null)` on both its success and its catch path (:185, :193), with no `isReevaluation` guard — the only thing that flag gates is the spinner at :161. So once a wrapper has settled empty and the token is set, the next re-evaluation nulls it, which is a real state change, and the wrapper re-renders with no class. Nothing restores it. The only non-null writer is `syncRenderedOutput` (:229-243), reached from the effect at :223-249 whose deps are `[hasRenderableChildren, renderToken]`; a re-evaluation that lands on the same branch verdict changes neither, because `setConditionsPassed` writes an equal boolean, `childrenToRender` is a memo on unchanged deps (:203-206), and `branchKey` (:209) is the same string, so `renderToken` keeps its identity and the effect does not re-run. The still-attached observer is registered `{ childList: true, subtree: true }` (:247) with no `attributes`, so React rewriting the wrapper's own class is invisible to it, and a subtree that already rendered to nothing produces no childList record either. Recovery needs a later real subtree mutation, which by definition is not coming while the branch stays empty. The triggers are ordinary guide traffic: `scheduleReevaluation` (:265-268) is wired at :300-310 to `interactive-action-completed` on both window and document, `datasources-changed`, `plugins-changed`, `popstate`, and `pathfinder:progress` step completions. With both classes gone the `<li>` is not `:empty` — it still holds the conditional div — so none of the three suppressors in src/styles/interactive.styles.ts (:363 `:empty`, :369, :449) matches and the base rule at :404-421 applies. In the never-occupied case that means an ordinal painted beside no content plus a consumed counter slot, so every later step in the section reads one higher than its authored position: the same harm as the round 2 blocker, which is why this keeps that finding's ID. In the previously-occupied case the slot is consumed by design, so the shift does not return, but the collapse does come off and the row reappears full-height with a number and nothing next to it — the state the round 1 retained-slot fix exists to prevent. The reachable shape is the one already named last round: a conditional whose branch holds parsed children that render to nothing, most concretely a nested conditional that settles empty. Two independent skeptics tried to refute this and could not. No test catches it: `retains an occupied numbering slot when re-evaluation empties the branch` exercises the separate `childrenToRender.length === 0` early return at :373-377, and `retains the outer slot when a populated nested conditional collapses` dispatches `interactive-action-completed` exactly once, at a moment when the outer wrapper still holds visible content and its token is already null, so the reset is a no-op there and the inner collapse's real child mutation sets the state afterwards.
   Required: Stop clearing `emptyRenderToken` inside `evaluateConditions` and let the DOM inspection be its only writer — the remount-on-authoring-change case the comment at :215-216 describes is already served by `renderToken` identity, so the two resets look redundant as well as harmful. If you would rather keep them, make the effect re-run after a re-evaluation commits, for example by adding `emptyRenderToken` to its deps or by also observing attributes on the wrapper. Either way, add a test that dispatches `interactive-action-completed` a second time after the wrapper has already settled empty and asserts the class and the collapsed computed style survive it.



PR Review: https://github.com/grafana/grafana-pathfinder-app/pull/1809
Summary: number conditional blocks in sections
Verdict: Request Changes
Results: 1 blocker, 0 non-blocking findings, 0 follow-ups

**Comment by dvd233** [comment]
Thanks for the Round 3 review. The current head `b3ca2d1da3f029af3be45a7049d1a7bdbdf1093c` includes the requested fix: `evaluateConditions` no longer clears `emptyRenderToken` on success or error, and the regression suite now dispatches a second same-branch re-evaluation while asserting the empty/retained marker and collapsed style remain. The focused interactive-conditional suite and full interactive-tutorial checks are green, with typecheck, lint, and Prettier clean. Could you re-review the current head?

**Review by moxious — APPROVED** [review]
Thanks for sticking with this through three rounds and landing the fix exactly where round 3 pointed: both `setEmptyRenderToken(null)` resets are gone from `evaluateConditions`, the renderToken-keyed DOM-inspection effect is now the only writer, and the new tests dispatch a second `interactive-action-completed` after the wrapper settles empty/retained to catch a repeat of this exact regression.

No blocking issues. This PR is mergeable.



PR Review: https://github.com/grafana/grafana-pathfinder-app/pull/1809
Summary: number conditional blocks in sections
Verdict: Approve
Results: 0 blockers, 0 non-blocking findings, 0 follow-ups

Once CI is green on this head, this is ready to merge.
